File: ring-android/app/src/main/java/cx/ring/fragments/ContactPickerFragment.java
package cx.ring.fragments;

import android.app.Dialog;
import android.os.Bundle;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.material.bottomsheet.BottomSheetBehavior;
import com.google.android.material.bottomsheet.BottomSheetDialog;
import com.google.android.material.bottomsheet.BottomSheetDialogFragment;
import com.google.android.material.chip.Chip;

import java.util.HashSet;
import java.util.Set;

import javax.inject.Inject;

import cx.ring.R;
import cx.ring.adapters.SmartListAdapter;
import cx.ring.client.HomeActivity;
import cx.ring.databinding.FragContactPickerBinding;
import net.jami.facades.ConversationFacade;
import net.jami.model.Contact;
import net.jami.smartlist.SmartListViewModel;
import cx.ring.viewholders.SmartListViewHolder;
import cx.ring.views.AvatarDrawable;
import dagger.hilt.android.AndroidEntryPoint;
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.disposables.CompositeDisposable;

@AndroidEntryPoint
public class ContactPickerFragment extends BottomSheetDialogFragment {

    public static final String TAG = ContactPickerFragment.class.getSimpleName();

    private FragContactPickerBinding binding = null;
    private SmartListAdapter adapter;
    private final CompositeDisposable mDisposableBag = new CompositeDisposable();

    private String mAccountId =  null;
    private final Set<Contact> mCurrentSelection = new HashSet<>();

    @Inject
    ConversationFacade mConversationFacade;

    public ContactPickerFragment() {
        // Required empty public constructor
    }

    public static ContactPickerFragment newInstance() {
        return new ContactPickerFragment();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        //setRetainInstance(true);
        //((JamiApplication) getActivity().getApplication()).getInjectionComponent().inject(this);
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);

    }

    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
        Dialog bdialog = super.onCreateDialog(savedInstanceState);
        if (bdialog instanceof BottomSheetDialog) {
            BottomSheetDialog dialog = (BottomSheetDialog) bdialog;
            BottomSheetBehavior<FrameLayout> behavior = dialog.getBehavior();
            behavior.setFitToContents(false);
            behavior.setSkipCollapsed(true);
            behavior.setState(BottomSheetBehavior.STATE_HALF_EXPANDED);
        }
        return bdialog;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        mDisposableBag.add(mConversationFacade.getContactList()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(conversations -> {
                    if (binding == null)
                        return;
                    adapter.update(conversations);
                }));
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragContactPickerBinding.inflate(getLayoutInflater(), container, false);
        adapter = new SmartListAdapter(null, new SmartListViewHolder.SmartListListeners() {
            @Override
            public void onItemClick(SmartListViewModel item) {
                mAccountId = item.getAccountId();

                boolean checked = !item.isChecked();
                item.setChecked(checked);
                adapter.update(item);

                Runnable remover = () -> {
                    mCurrentSelection.remove(item.getContacts().get(0));
                    if (mCurrentSelection.isEmpty())
                        binding.createGroupBtn.setEnabled(false);
                    item.setChecked(false);
                    adapter.update(item);
                    View v = binding.selectedContacts.findViewWithTag(item);
                    if (v != null)
                        binding.selectedContacts.removeView(v);
                };

                if (checked) {
                    if (mCurrentSelection.add(item.getContacts().get(0))) {
                        Chip chip = new Chip(requireContext(), null, R.style.Widget_MaterialComponents_Chip_Entry);
                        chip.setText(item.getContactName());
                        chip.setChipIcon(new AvatarDrawable.Builder()
                                .withViewModel(item)
                                .withCircleCrop(true)
                                .withCheck(false)
                                .build(binding.getRoot().getContext()));
                        chip.setCloseIconVisible(true);
                        chip.setTag(item);
                        chip.setOnCloseIconClickListener(v -> remover.run());
                        binding.selectedContacts.addView(chip);
                    }
                    binding.createGroupBtn.setEnabled(true);
                } else {
                    remover.run();
                }
            }

            @Override
            public void onItemLongClick(SmartListViewModel item) {

            }
        }, mDisposableBag);
        binding.createGroupBtn.setOnClickListener(v -> mDisposableBag.add(mConversationFacade.createConversation(mAccountId, mCurrentSelection)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(conversation -> {
                    ((HomeActivity) requireActivity()).startConversation(mAccountId, conversation.getUri());
                    Dialog dialog = getDialog();
                    if (dialog != null)
                        dialog.cancel();
                })));
        binding.contactList.setAdapter(adapter);
        return binding.getRoot();
    }

    @Override
    public void onDestroyView() {
        mDisposableBag.clear();
        binding = null;
        adapter = null;
        super.onDestroyView();
    }

    @Override
    public int getTheme() {
        return R.style.BottomSheetDialogTheme;
    }
}

File: ring-android/app/src/main/java/cx/ring/fragments/ContactPickerFragment.kt
package cx.ring.fragments

import android.app.Dialog
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import com.google.android.material.chip.Chip
import cx.ring.R
import cx.ring.adapters.ContactPickerAdapter
import cx.ring.client.HomeActivity
import cx.ring.databinding.FragContactPickerBinding
import cx.ring.viewholders.ContactPickerViewHolder.ContactPickerListeners
import cx.ring.views.AvatarDrawable
import dagger.hilt.android.AndroidEntryPoint
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.disposables.CompositeDisposable
import net.jami.model.Contact
import net.jami.model.Conversation
import net.jami.services.ConversationFacade
import net.jami.smartlist.ConversationItemViewModel
import javax.inject.Inject

@AndroidEntryPoint
class ContactPickerFragment : BottomSheetDialogFragment() {
    private var binding: FragContactPickerBinding? = null
    private var adapter: ContactPickerAdapter? = null
    private val mDisposableBag = CompositeDisposable()
    private var mAccountId: String? = null
    private val mCurrentSelection: MutableSet<Contact> = HashSet()

    @Inject
    lateinit var mConversationFacade: ConversationFacade

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val bdialog = super.onCreateDialog(savedInstanceState)
        if (bdialog is BottomSheetDialog) {
            val behavior = bdialog.behavior
            behavior.isFitToContents = false
            behavior.skipCollapsed = true
            behavior.state = BottomSheetBehavior.STATE_HALF_EXPANDED
        }
        return bdialog
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        mDisposableBag.add(mConversationFacade.getConversationList()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({ conversations ->
                val contacts = conversations.filter { it.mode == Conversation.Mode.Legacy || it.mode == Conversation.Mode.OneToOne }
                adapter?.update(contacts)
            }){ e -> Log.e(TAG, "No contact to create a group!", e) })
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = FragContactPickerBinding.inflate(layoutInflater, container, false)
        adapter = ContactPickerAdapter(null, object : ContactPickerListeners {
            override fun onItemClick(item: ConversationItemViewModel) {
                mAccountId = item.accountId
                val checked = !item.isChecked
                item.isChecked = checked
                adapter!!.update(item)
                val contact = item.getContact() ?: return
                val remover = Runnable {
                    mCurrentSelection.remove(contact.contact)
                    if (mCurrentSelection.isEmpty()) binding!!.createGroupBtn.isEnabled = false
                    item.isChecked = false
                    adapter!!.update(item)
                    val v = binding!!.selectedContacts.findViewWithTag<View>(item)
                    if (v != null) binding!!.selectedContacts.removeView(v)
                }
                if (checked) {
                    if (mCurrentSelection.add(contact.contact)) {
                        val chip = Chip(requireContext(), null, R.style.Widget_MaterialComponents_Chip_Entry).apply {
                            text = item.title
                            chipIcon = AvatarDrawable.Builder()
                                .withContact(contact)
                                .withCircleCrop(true)
                                .withCheck(false)
                                .withPresence(false)
                                .build(binding!!.root.context)
                            isCloseIconVisible = true
                            tag = item
                            setOnCloseIconClickListener { remover.run() }
                        }
                        binding!!.selectedContacts.addView(chip)
                    }
                    binding!!.createGroupBtn.isEnabled = true
                } else {
                    remover.run()
                }
            }

            override fun onItemLongClick(item: ConversationItemViewModel) {}
        })
        binding!!.createGroupBtn.setOnClickListener { v: View? ->
            mDisposableBag.add(mConversationFacade.createConversation(mAccountId!!, mCurrentSelection)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { conversation: Conversation ->
                    (requireActivity() as HomeActivity).startConversation(conversation.accountId, conversation.uri)
                    val dialog = dialog
                    dialog?.cancel()
                })
        }
        binding!!.contactList.adapter = adapter
        return binding!!.root
    }

    override fun onDestroyView() {
        mDisposableBag.clear()
        binding = null
        adapter = null
        super.onDestroyView()
    }

    override fun getTheme(): Int {
        return R.style.BottomSheetDialogTheme
    }

    companion object {
        val TAG = ContactPickerFragment::class.java.simpleName
    }
}